docs/eventing/samples/container-source/README.md

```diff
@@ -0,0 +1,133 @@
+- [`docker`](https://docs.docker.com/install/) to be installed
+
+### Create a Knative Service
+
+In order to verify `ContainerSource` is working, we will create a Event Display Service that dumps incoming messages to its log.
+
+```yaml
+apiVersion: serving.knative.dev/v1alpha1
+kind: Service
+metadata:
+  name: event-display
+spec:
+  runLatest:
+    configuration:
+      revisionTemplate:
+        spec:
+          container:
+            image: gcr.io/knative-releases/github.com/knative/eventing-sources/cmd/event_display
+```
+
+Use following command to create the service from `service.yaml`:
+
+```shell
+kubectl apply --filename service.yaml
+```
+
+### Create a ContainerSource using the heartbeats image
+
+In order to run the heartbeats container as an event source, you have to create a concrete ContainerSource with specific arguments and environment settings. Be sure to replace
+`heartbeats_image_uri` with a valid uri for your heartbeats image in your image repo. Note that arguments and environment variables are set and will be passed to the container.
+
+```yaml
+apiVersion: sources.eventing.knative.dev/v1alpha1
+kind: ContainerSource
+metadata:
+  name: test-heartbeats
+spec:
+  image: <heartbeats_image_uri>
+  sink:
+    apiVersion: serving.knative.dev/v1alpha1
+    kind: Service
+    name: event-display
+  args: 
+    - --period=1
+  env:
+    - name: POD_NAME
+      value: "mypod"
+    - name: POD_NAMESPACE
+      value: "event-test"
+```
+
+Use the following command to create the event source from `heartbeats-source.yaml`:
+
+```shell
+kubectl apply --filename heartbeats-source.yaml
+```
+
+### Verify
+
+We will verify that the message was sent to the Knative eventing system by
+looking at message dumper logs.
+
+```shell
+kubectl logs -l serving.knative.dev/service=event-display -c user-container --since=10m
+```
+
+You should see log lines showing the request headers and body from the source:
+
+```
+☁️  CloudEvent: valid ✅
+Context Attributes,
+  SpecVersion: 0.2
+  Type: dev.knative.eventing.samples.heartbeat
+  Source: https://github.com/knative/eventing-sources/cmd/heartbeats/#event-test/mypod
+  ID: cd1f5f24-12dd-489d-aff4-23302c6091fa
+  Time: 2019-04-04T08:38:24.833521851Z
+  ContentType: application/json
+  Extensions:
+    beats: true
+    heart: yes
+    the: 42
+Transport Context,
+  URI: /
+  Host: event-display.default.svc.cluster.local
+  Method: POST
+Data,
+  {
+    "id": 2,
+    "label": ""
+  }
+```
+
+## Create a new event source using ContainerSource
+
+In order to create a new event source using ContainerSource, you will create a container image at first, and then create a ContainerSource with the image uri and specify the values of parameters.
+
+### Develop, build and publish a container image
+
+The container image can be developed with any language, build and publish with any tools you like. Here are some basic guidelines:
+- The container image must have a `main` method to start with.
```

Comment:
in every language ?

Reply:
yes. Actually, ContainerSource only needs a URI of a docker image in image repo.